Finished BJS2-1501 task (Follow User) feature. Finished BJS2-1504 task (Unfollow User) feature.

starkoracia opened the pull request:
Finished BJS2-1501 task (Follow User) feature.
Finished BJS2-1504 task (Unfollow User) feature.
feature {
+ Created classes: DataValidationException, GlobalExceptionHandler, SubscriptionControllerImpl, SubscriptionServiceImpl, SubscriptionService.
}
+ feature(SubscriptionServiceImpl) {
+ created unfollowUser() method
  }
  feature(SubscriptionControllerImpl) {
+ created unfollowUser() method
  }
  refactor() {
  -+ (SubscriptionServiceImplTest) moved validation logic to a separate method
  -+ (SubscriptionControllerImpl) renamed some names
  }
  test {
+ Created classes: SubscriptionServiceImplTest, SubscriptionControllerImplTest
  }

starkoracia committed:
Finished BJS2-1501 task (Follow User) feature.
feature {
 + Created classes: DataValidationException, GlobalExceptionHandler, SubscriptionControllerImpl, SubscriptionServiceImpl, SubscriptionService.
}

starkoracia committed:
Finished BJS2-1504 task (Unfollow User) feature.
feature(SubscriptionServiceImpl) {
 + created unfollowUser() method
}
feature(SubscriptionControllerImpl) {
 + created unfollowUser() method
}
refactor() {
 -+ (SubscriptionServiceImplTest) moved validation logic to a separate method
 -+ (SubscriptionControllerImpl) renamed some names
}
test {
 + Created classes: SubscriptionServiceImplTest, SubscriptionControllerImplTest
}

PavelAP96 commented on src/main/java/school/faang/user_service/controller/subscription/SubscriptionControllerImpl.java:
```diff
@@ -0,0 +1,46 @@
+@RestController
+@RequestMapping("subscriptions")
+@RequiredArgsConstructor
+public class SubscriptionControllerImpl {
```
К названию класса обычно добавляют Impl, если они реализуют как - то интерфейс. В данном случае в этом нет необходимости, обычно в проектах контроллеры называют без таких преписок, здесь будет просто SubscriptionController

starkoracia replied:
++ Втыканул. ))

PavelAP96 commented on src/main/java/school/faang/user_service/controller/subscription/SubscriptionControllerImpl.java:
```diff
@@ -0,0 +1,37 @@
+        if (followerId == followeeId) {
+            throw new DataValidationException("User can not be follow to himself");
+        }
+    }
```
понимаю, что в задании было написано сделать этот тут, но я все - таки советую перенести эту логику в SubscriptionService.
Потому что это все - таки некая бизнес - логика нашего приложения, и она должна быть на уровне сервиса.

starkoracia replied:
++ Согласен, хотел так и делать пока Инна не писала в общем чате )) Перенесу в сервис

PavelAP96 commented on src/main/java/school/faang/user_service/controller/subscription/SubscriptionControllerImpl.java:
```diff
@@ -0,0 +1,37 @@
+    private final SubscriptionService subscriptionService;
+
+    @PostMapping("/follow")
+    public Map.Entry<String, Boolean> followUser(@PathParam("followerId") long followerId,
```
по заданию метод должен возвращать void
можешь мне прям здесь в комменте ответить, зачем возвращать Map.Entry<String, Boolean>?
уже пытался через Postman тестить?

starkoracia replied:
![image](https://github.com/CorporationX/user_service/assets/49105510/536893f5-38bd-4adb-95cb-3b157f0b5095)
Хотел чтобы возвращало true. Хотел иметь минимальный json обьект вместо просто true значения.
До все тестил через Postman все гуд.

PavelAP96 replied:
понял
пока пойдет)

PavelAP96 commented on src/test/java/school/faang/user_service/controller/subscription/SubscriptionControllerImplTest.java:
```diff
@@ -0,0 +1,73 @@
+    SubscriptionControllerImpl subscriptionController;
+
+    @Test
+    void should_ReturnTrue_when_FollowIsSuccessful() {
```
в названиях используем camelCase, я бы не советовал юзать нижние подчеркивания

starkoracia replied:
++ Оки.

PavelAP96 commented on src/main/java/school/faang/user_service/controller/subscription/SubscriptionControllerImpl.java:
```diff
@@ -0,0 +1,46 @@
+        return Map.entry("isUnfollowed", true);
+    }
+
+    private void validateUserIds(long followerId, long followeeId) {
+        if (followerId <= 0 || followeeId <= 0) {
+            throw new DataValidationException("User identifiers must be positive numbers");
+        }
+        if (followerId == followeeId) {
+            throw new DataValidationException("Follower and followee the same user");
+        }
+    }
```
понимаю, что в задании было написано сделать этот тут, но я все - таки советую перенести эту логику в SubscriptionService.
Потому что это все - таки некая бизнес - логика нашего приложения, и она должна быть на уровне сервиса.

starkoracia replied:
++ Согласен, хотел так и делать пока Инна не писала в общем чате )) Перенесу в сервис

PavelAP96 commented on src/test/java/school/faang/user_service/controller/subscription/SubscriptionControllerImplTest.java:
```diff
@@ -0,0 +1,124 @@
+        long followerId = 5L;
+        long followeeId = 2L;
+
+        Map.Entry<String, Boolean> result = subscriptionController.unfollowUser(followerId, followeeId);
+
+        assertNotNull(result);
+        assertEquals("isUnfollowed", result.getKey());
+        assertTrue(result.getValue());
```
почему бы все это ни написать в 1 строчку:
`assertEquals( Map.entry("isUnfollowed", true), subscriptionController.unfollowUser(followerId, followeeId));`

это же касается и теста для подписки

starkoracia replied:
Показалось более читабельно. Меняю на одну строчку )

PavelAP96 commented on src/test/java/school/faang/user_service/controller/subscription/SubscriptionControllerImplTest.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    @Test
+    void should_ThrowException_when_UnfollowUserIdsAreInvalid() {
+        long invalidFollowerId = -1L;
+        long invalidFolloweeId = -2L;
+        String expectedExMessage = "User identifiers must be positive numbers";
+
+        Exception exceptionForFollower = assertThrows(
+                DataValidationException.class,
+                () -> subscriptionController.unfollowUser(invalidFollowerId, 1L));
+        assertEquals(expectedExMessage, exceptionForFollower.getMessage());
+
+        Exception exceptionForFollowee = assertThrows(
+                DataValidationException.class,
+                () -> subscriptionController.unfollowUser(1L, invalidFolloweeId)
+        );
+        assertEquals(expectedExMessage, exceptionForFollowee.getMessage());
+
+        verify(subscriptionService, never()).unfollowUser(anyLong(), anyLong());
```
обычно не проверяют, какое сообщение в исключении
достаточно проверить, что исключение выброшено

это же касается для метода тестирования подписки

starkoracia replied:
++ Ок) Так еще проще)

PavelAP96 commented on src/test/java/school/faang/user_service/controller/subscription/SubscriptionControllerImplTest.java:
```diff
@@ -0,0 +1,124 @@
+    @Test
+    void should_ThrowException_when_UnfollowUserIdsAreTheSame() {
+        long sameUserId = 1L;
+        String expectedExMessage = "Follower and followee the same user";
+
+        Exception exception = assertThrows(
+                DataValidationException.class,
+                () -> subscriptionController.unfollowUser(sameUserId, sameUserId)
+        );
+        assertEquals(expectedExMessage, exception.getMessage());
+
+        verify(subscriptionService, never()).unfollowUser(anyLong(), anyLong());
```
обычно не проверяют, какое сообщение в исключении
достаточно проверить, что исключение выброшено

это же касается для метода тестирования подписки

starkoracia replied:
++

PavelAP96 commented on src/test/java/school/faang/user_service/service/subscription/SubscriptionServiceImplTest.java:
```diff
@@ -0,0 +1,64 @@
+    private SubscriptionServiceImpl subscriptionService;
+
+    @Test
+    void should_DeleteSubscription_when_FollowerAndFolloweeIdsAreDifferent() {
```
в названиях используем camelCase, я бы не советовал юзать нижние подчеркивания

PavelAP96 commented on src/test/java/school/faang/user_service/service/subscription/SubscriptionServiceImplTest.java:
```diff
@@ -0,0 +1,64 @@
+        when(subscriptionRepository.existsByFollowerIdAndFolloweeId(followerId, followeeId))
+                .thenReturn(Boolean.TRUE);
+
+        DataValidationException exception = assertThrows(
+                DataValidationException.class,
+                () -> subscriptionService.followUser(followerId, followeeId));
+
+        assertEquals("This subscription already exists", exception.getMessage());
```
достаточно проверить, что исключение выброшено

starkoracia replied:
++

starkoracia committed:
BJS2-1414 task refactor.

refactor() {
 -+ SubscriptionControllerImpl was renamed as SubscriptionController and the test class as well
 -+ Moved validation to SubscriptionServiceImpl
}

starkoracia committed:
BJS2-1414 test refactor.

refactor() {
 -+ Was renamed test's names to camelCase
 -+ Was removed exception's messages check
 -+ Moved validations to the service class

PavelAP96 reviewed (approved):
я бы только вынес объявление переменных в тестах в блок @BeforeEach
повторяются
